core/compiler.common/src/org/jetbrains/kotlin/types/AbstractTypeChecker.kt

```diff
@@ -404,7 +404,9 @@ object AbstractTypeChecker {
 
         if (subType.isStubType() || superType.isStubType()) return isStubTypeEqualsToAnything
 
-        val superTypeCaptured = superType.asCapturedType()
+        // superType might be a definitely notNull type (see KT-42824)
+        val superOriginalType = superType.asDefinitelyNotNullType()?.original() ?: superType
```

Comment:
I believe we should propagate `!!`-nullability inside `lowerType`, similar to what you did in https://github.com/JetBrains/kotlin/pull/3869. So, subtyping check `T <: Captured(in Foo)!!` will be transformed into `T <: Foo!!`, otherwise we'll just lose information about nullability here